awsauth/sign.go

```diff
@@ -6,46 +6,71 @@ import (
 	"net/http"
 	"time"
 
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
 	signerV4 "github.com/aws/aws-sdk-go/aws/signer/v4"
 	awsauth "github.com/smartystreets/go-aws-auth"
 )
 
 type Signer struct {
-	awsRegion    string
-	awsSDKSigner bool
-	awsService   string
+	awsRegion  string
+	awsService string
+	v4         *signerV4.Signer
 }
 
-func NewSigner(awsSDKSigner bool, awsRegion, awsService string) *Signer {
-	return &Signer{
-		awsRegion:    awsRegion,
-		awsSDKSigner: awsSDKSigner,
-		awsService:   awsService,
+func NewAwsSigner(awsFilename, awsProfile, awsRegion, awsService string) (signer *Signer, err error) {
+	if err = validateAwsSDKSigner(awsRegion, awsService); err != nil {
+		return
 	}
+
+	var sess *session.Session
+	sess, err = session.NewSession()
+	if err != nil {
+		return
+	}
+
+	creds := credentials.NewChainCredentials(
+		[]credentials.Provider{
+			&credentials.EnvProvider{},
+			&credentials.SharedCredentialsProvider{
+				Filename: awsFilename,
+				Profile:  awsProfile,
+			},
+			&ec2rolecreds.EC2RoleProvider{
+				Client: ec2metadata.New(sess),
+			},
+		})
+
```

Comment:
Small thing but bracket should probably closed on same column as it was opened (on next line)